### DISCUSSION TITLE
How to add additional validation for a model on create/update?

### DISCUSSION THREAD

**Discussion by onecrayon** [strapi/strapi#8101] [Legacy Topics]
Is there a place that I can hook logic in to add additional validation for a model? I see that a [similar question](https://github.com/strapi/strapi/discussions/6787) was asked a while back and got no response, but I wanted to ask again with details about my particular setup:

I have a table of "items" that include various relationships (images, PDFs, custom usage instructions, etc.). Each item has a self-referential relationship like `Item.reimplements_item`. If this field is populated, then when looking up related content for this item, the results should be for the item in `reimplements_item` (logic here is that items get refreshed potentially yearly, but the actual related instructions don't change).

This all works great when querying, because I can filter for that stuff in the endpoints via controller logic.

However, I want to ensure that if an item has `reimplements_item` populated, nothing else about it can be modified (and that it is not possible to reference an item in `reimplements_item` if that item also has a populated `reimplements_item`; they all need to point to the same root item, or I get stuck with arbitrary length chains). To do this, I need to hook into the creation and update logic somehow and ensure that I throw a validation error when making changes that break these rules.

Is this possible? I can shift this logic over to the front-end, but I would **really** like to restrict it on the Strapi side, because every time I've trusted data integrity to front-end validation, I've gotten burned eventually (plus if it's in Strapi, then presumably that would also affect the admin editing app). 😅

**Comment by derrickmehaffy** [comment]
You will most likely want to take a look at the entityValidator function (we just recently added this to our docs, https://github.com/strapi/strapi/pull/7999). 

https://github.com/strapi/strapi/blob/master/packages/strapi/lib/services/entity-validator/index.js

**Reply by onecrayon** [reply]
@derrickmehaffy Hm, okay. Where in the docs is that, exactly? Would I need to create a custom [service](https://strapi.io/documentation/v3.x/concepts/services.html#concept) for my model, call the entity validator, and then layer in my own custom validation logic afterward (using the exception pattern in the file you linked above)? Or can I hook into that system at a lower level to extend the available validation logic without needing to overwrite the service?

**Reply by derrickmehaffy** [reply]
It isn't in the docs yet (see that PR I linked), as for the entity validator that is applied automatically but if you need more in depth validation then yes you will want to customize the service, you will be able to return errors to the user using something like `return ctx.badRequest('Some message here') or any of the other supported boom responses: 

https://github.com/strapi/strapi/blob/a22baf9a0d74c0ebf3012bc56f470bed08be20bc/packages/strapi/lib/middlewares/boom/index.js#L12-L43

Due to where that function is located you won't be able to hook into it and extend it without forking the `strapi` package.

**Reply by onecrayon** [reply]
@derrickmehaffy How do I access `ctx` from within services? I tried `async create(data, { ctx, files } = {}) {` but it was undefined.

**Reply by derrickmehaffy** [reply]
Ah my apologies you are correct, the `ctx` isn't passed into services, you'll need to handle the check and the `return ctx.badRequest()` at the controller level.

**Reply by onecrayon** [reply]
@derrickmehaffy I'd really like to keep this at the services level so that I can be sure that the restrictions will be applied when people try to edit things through the Strapi admin. If I raise an error in a service, how would that percolate in the Strapi admin? Would it capture and report it, or be likely to break it in unexpected ways?

**Reply by derrickmehaffy** [reply]
It would not be handled by the Admin as the admin has it's own internal API that doesn't use the `controllers` or `services` Specifically this here: 

https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-content-manager/controllers/ContentManager.js
https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-content-manager/services/ContentManager.js

The only thing the Admin and REST/GraphQL share in common are the queries (executed by the services) and the lifecycle callbacks (executed by the queries).

The Lifecycle callbacks do share a similar scope however we are aware of a feature request to pass the ctx into it, but to my knowledge you won't be able to "stop the process, and return an error" from them. 

I think this is a larger topic that we need to have on the https://forum.strapi.io/c/discussions/6 and loop in the backend engineers to come up a viable solution going forward (I certainly understand the pain point)

**Reply by derrickmehaffy** [reply]
(We plan to get rid of GitHub discussions sometime around Q1 2021 and I don't want to loose this discussion here when we do)

**Reply by onecrayon** [reply]
Moved it over: https://forum.strapi.io/t/custom-model-validation-across-all-layers/236

**Reply by derrickmehaffy** [reply]
Locking discussion per forum thread
